src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/ChainPal.cs

```diff
@@ -120,25 +128,66 @@ private SafeCreateHandle PreparePoliciesArray(bool checkRevocation)
 
-        private SafeCreateHandle PrepareCertsArray(ICertificatePal cert, X509Certificate2Collection extraStore)
+        private SafeCreateHandle PrepareCertsArray(
+            ICertificatePal cert,
+            X509Certificate2Collection extraStore,
+            X509Certificate2Collection customTrustStore,
+            X509ChainTrustMode trustMode)
         {
-            IntPtr[] ptrs = new IntPtr[1 + (extraStore?.Count ?? 0)];
-            SafeHandle[] safeHandles = new SafeHandle[ptrs.Length];
-
+            int certificateCount = 1 + (extraStore?.Count ?? 0) + (trustMode == X509ChainTrustMode.CustomRootTrust ? customTrustStore?.Count ?? 0 : 0);
+            int certificatesAdded = 1;
+            SafeHandle[] safeHandles = new SafeHandle[certificateCount];
             AppleCertificatePal applePal = (AppleCertificatePal)cert;
-
             safeHandles[0] = applePal.CertificateHandle;
 
             if (extraStore != null)
             {
-                for (int i = 0; i < extraStore.Count; i++)
+                foreach (X509Certificate2 extraCert in extraStore)
                 {
-                    AppleCertificatePal extraCertPal = (AppleCertificatePal)extraStore[i].Pal;
+                    AppleCertificatePal extraCertPal = (AppleCertificatePal)extraCert.Pal;
+                    safeHandles[certificatesAdded] = extraCertPal.CertificateHandle;
+                    certificatesAdded++;
+                }
+            }
 
-                    safeHandles[i + 1] = extraCertPal.CertificateHandle;
+            if (trustMode == X509ChainTrustMode.CustomRootTrust && customTrustStore != null)
+            {
+                for (int i = 0; i < customTrustStore.Count; i++)
+                {
+                    AppleCertificatePal extraCertPal = (AppleCertificatePal)customTrustStore[i].Pal;
+                    safeHandles[certificatesAdded + i] = extraCertPal.CertificateHandle;
                 }
             }
 
+            return GetCertsArray(safeHandles);
+        }
+
+        private SafeCreateHandle PrepareCustomCertsArray(X509Certificate2Collection customTrustStore)
+        {
+            X509Certificate2Collection rootCertificates = new X509Certificate2Collection();
+            SafeHandle[] safeHandles;
+
+            foreach (X509Certificate2 cert in customTrustStore)
+            {
+                if (cert.SubjectName.RawData.ContentsEqual(cert.IssuerName.RawData))
+                {
+                    rootCertificates.Add(cert);
+                }
+            }
+
+            safeHandles = new SafeHandle[rootCertificates.Count];
+            for (int i = 0; i < rootCertificates.Count; i++)
+            {
+                AppleCertificatePal extraCertPal = (AppleCertificatePal)rootCertificates[i].Pal;
+                safeHandles[i] = extraCertPal.CertificateHandle;
+            }
+
+            return GetCertsArray(safeHandles);
+        }
+
+        private SafeCreateHandle GetCertsArray(SafeHandle[] safeHandles)
```

Comment:
The implementation of this method makes me a little nervous, and has some duplicated code.  It's pre-existing your PR, but either here or subsequently it'd be nice to see it cleaned up.  It could be written more like (haven't compiled or tested this):
```C#
        private SafeCreateHandle GetCertsArray(SafeHandle[] safeHandles)
        {
            int idx = 0;
            try
            {
                var ptrs = new IntPtr[safeHandles.Length];
                for (; idx < safeHandles.Length; idx++)
                {
                    SafeHandle handle = safeHandles[idx];
                    bool addedRef = false;
                    handle.DangerousAddRef(ref addedRef);
                    ptrs[idx] = handle.DangerousGetHandle();
                }

                // Creating the array has the effect of calling CFRetain() on all of the pointers, so the native
                // resource is safe even when we DangerousRelease=>ReleaseHandle them.
                SafeCreateHandle certsArray = Interop.CoreFoundation.CFArrayCreate(ptrs, (UIntPtr)ptrs.Length);
                _extraHandles.Push(certsArray);
                return certsArray;
            }
            finally
            {
                for (idx--; idx >= 0; idx--)
                {
                    safeHandles[idx].DangerousRelease();
                }
            }
        }
```